support/k8s/k8s-workload-registrar/controllers/pod_controller.go

```diff
@@ -0,0 +1,313 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type PodReconcilerMode int32
+
+const (
+	PodReconcilerModeServiceAccount PodReconcilerMode = iota
+	PodReconcilerModeLabel
+	PodReconcilerModeAnnotation
+)
+
+// PodReconciler reconciles a Pod object
+type PodReconciler struct {
+	client.Client
+	TrustDomain    string
+	Mode           PodReconcilerMode
+	Value          string
+	RootID         string
+	SpireClient    registration.RegistrationClient
+	ClusterDNSZone string
+	AddPodDNSNames bool
+}
+
+type WorkloadSelectorSubType string
+
+const (
+	PodNamespaceSelector WorkloadSelectorSubType = "ns"
+	PodNameSelector      WorkloadSelectorSubType = "pod-name"
+)
+
+const endpointSubsetAddressReferenceField string = ".subsets.addresses.targetRef.uid"
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch
+
+func (r *PodReconciler) k8sWorkloadSelector(selector WorkloadSelectorSubType, value string) *common.Selector {
+	return &common.Selector{
+		Type:  "k8s",
+		Value: fmt.Sprintf("%s:%s", selector, value),
+	}
+}
+
+func (r *PodReconciler) selectorsToNamespacedName(selectors []*common.Selector) *types.NamespacedName {
+	podNamespace := ""
+	podName := ""
+	for _, selector := range selectors {
+		if selector.Type == "k8s" {
+			splitted := strings.SplitN(selector.Value, ":", 2)
+			if len(splitted) > 1 {
+				switch WorkloadSelectorSubType(splitted[0]) {
+				case PodNamespaceSelector:
+					podNamespace = splitted[1]
+				case PodNameSelector:
+					podName = splitted[1]
+				}
+			}
+		}
+	}
+	if podNamespace != "" && podName != "" {
+		return &types.NamespacedName{
+			Namespace: podNamespace,
+			Name:      podName,
+		}
+	}
+	return nil
+}
+
+func (r *PodReconciler) makeSpiffeID(obj ObjectWithMetadata) string {
+	return r.makeSpiffeIDForPod(obj.(*corev1.Pod))
+}
+
+func (r *PodReconciler) mungeIP(ip string) string {
+	if strings.Contains(ip, ".") {
+		// IPv4
+		return strings.Replace(ip, ".", "-", -1)
+	}
+	if strings.Contains(ip, ":") {
+		// IPv6
+		return strings.Replace(ip, ":", "-", -1)
+	}
+	return ip
+}
+
+func (r *PodReconciler) fillEntryForPod(ctx context.Context, entry *common.RegistrationEntry, pod *corev1.Pod) (*common.RegistrationEntry, error) {
+	if !r.AddPodDNSNames {
+		return entry, nil
+	}
+
+	names := make(map[string]bool)
+
+	endpointsList := corev1.EndpointsList{}
+	if err := r.List(ctx, &endpointsList, client.InNamespace(pod.Namespace), client.MatchingFields{endpointSubsetAddressReferenceField: pod.Name}); err != nil {
+		return nil, err
+	}
+
+	for _, endpoints := range endpointsList.Items {
+		endpoints := endpoints
+		// Based on https://github.com/kubernetes/dns/blob/master/docs/specification.md
+		// We cheat slightly and don't check the type of service (headless or not), we just add all possible names.
+
+		// 2.3.1 and 2.4.1: <service>.<ns>.svc.<zone>
+		names[fmt.Sprintf("%s.%s.svc.%s", endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
```

Comment:
Nope, since the first DNS is put into the `CN` of the SubjectName for X509-SVIDs, it serves as alternative authC for consumers that are not (yet) able to read the spiffe ID from the SAN URI (eg LDAP). Since we want to convey the identity, `<hostname>.<service>.<ns>.svc.<zone>` is the most specific and should be the first item to transcend into the `CN` attribute

Reply:
I wasn't aware of that CN behaviour. I don't think we should use the .svc. address forms, as there can be multiple k8s Services that reference the same Pod (note that endpoints is a loop variable): which one would you choose? After looking a bit more carefully I realised I'd missed the `pod-ip-address.my-namespace.pod.cluster-domain.example` case entirely, which is the most unique likely valid name for the pod. I've added that, and made sure it's the first in the list.

I wonder if this is still wrong though. If it's expected for some consumers to use it in place of the spiffe ID, then it should probably be generated based on a similar mechanism to the spiffe ID itself? Something like converting the generated SpiffeId by taking its path components, separating with . instead of /, reversing, and sticking on the front of the trust domain: e.g. given `spiffe://foo.trust.domain/bar/baz` the CN would be `baz.bar.foo.trust.domain`. The name wont be in any way a valid DNS entry, but it'll be more true to the intent perhaps.

Reply:
> I wonder if this is still wrong though. If it's expected for some consumers to use it in place of the spiffe ID, then it should probably be generated based on a similar mechanism to the spiffe ID itself?

I'm pondering about this as well!

I think, the `CN` attribute perfectly can host the spiffe ID _at is_. By no means does it need to be a DNS name., at all.

@azdagron Can you enlighten us on why the first DNS is used for the CN attribute instead of, say, the plain Spiffe ID string.

In my use case, now I have to handle two strings that refer to the same identity. And with @JonathanO 's ipmlementation it just gets more complicated, since eventually the several pod-ip addresses can refer to the same identity.

I'd argue, by default the spiffe ID itself shall be put into the `CN` attribute.

Reply:
Let's continue this discussion on in the issue. Thanks for bringing this up!